# Reading and using BWA

##Preprocessing

The command !apt-get install -y bwa installs the BWA (Burrows-Wheeler Aligner) software on your system, enabling you to perform sequence alignment of reads against a reference genome.

In [ ]:
# Install BWA
!apt-get install -y bwa

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Suggested packages:
  samtools
The following NEW packages will be installed:
  bwa
0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.
Need to get 195 kB of archives.
After this operation, 466 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bwa amd64 0.7.17-6 [195 kB]
Fetched 195 kB in 1s (134 kB/s)
Selecting previously unselected package bwa.
(Reading database ... 123621 files and directories currently installed.)
Preparing to unpack .../bwa_0.7.17-6_amd64.deb ...
Unpacking bwa (0.7.17-6) ...
Setting up bwa (0.7.17-6) ...
Processing triggers for man-db (2.10.2-1) ...


Unzips the Reference file

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Unzip the uploaded file
!unzip "/content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip" -d "/content/"

Archive:  /content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/mycobacterium_tuberculosis_h37rv_2_data.zip
  inflating: /content/mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta  
  inflating: /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta  
  inflating: /content/mycobacterium_tuberculosis_h37rv_2_contigs.agp  


In [ ]:
# List the files in the current directory
!ls -lh

total 8.6M
drwx------ 6 root root 4.0K Oct 11 14:05 drive
-rw-r--r-- 1 root root   53 Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_contigs.agp
-rw-r--r-- 1 root root 4.3M Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_contigs.fasta
-rw-r--r-- 1 root root 4.3M Jun  9  2014 mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta
drwxr-xr-x 1 root root 4.0K Oct  8 13:26 sample_data


The output of the `!ls -lh` command lists the files and directories in the current working directory with detailed information. Here's a breakdown of each line:

 General Information

-   **Total 12MB**: The total size of all files in the current directory is 12 Megabytes.

 Summary of Differences

| **Feature** | **Contigs (`contigs.fasta`)** | **Supercontigs (`supercontigs.fasta`)** |
| --- | --- | --- |
| **Definition** | Continuous sequences assembled from reads | Linked contigs with known order/orientation |
| **Length** | Shorter | Longer |
| **Gaps** | No gaps | May contain gaps (`N` bases) |
| **Assembly Level** | Initial stage of assembly | Higher level, linked contigs (scaffolds) |
| **Use Cases** | Local sequence analysis | Genome structure and gene prediction |
| **Completeness** | Less complete representation of genome | More complete, near full-chromosome length |

Installing both `bwa` and `samtools` together allows you to perform a complete workflow for analyzing sequencing data:

1.  **Align Reads**: Use `bwa` to align your sequencing reads to a reference genome.
2.  **Process Alignments**:
    -   Convert the resulting SAM file to BAM format using `samtools`.
    -   Sort and index the BAM file for efficient access and analysis.
    -   Use `samtools` to generate summary statistics, view alignments, and prepare data for variant calling or visualization.

-   **Indexing**: Preprocessing the reference genome to create data structures (BWT, SA) that enable fast searching and alignment.
-   **Example**: A read is aligned to the reference using the index, which helps BWA quickly find the correct position without scanning the entire genome.
-   **Efficiency**: Indexing significantly speeds up the alignment process by reducing the number of comparisons and memory usage.

In [ ]:
# Install BWA and samtools
!apt-get install -y bwa samtools

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
bwa is already the newest version (0.7.17-6).
The following additional packages will be installed:
  libhts3 libhtscodecs2
Suggested packages:
  cwltool
The following NEW packages will be installed:
  libhts3 libhtscodecs2 samtools
0 upgraded, 3 newly installed, 0 to remove and 49 not upgraded.
Need to get 963 kB of archives.
After this operation, 2,270 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]
Fetched 963 kB in 2s (506 kB/s)
Selecting previously unselected package libhtscodecs2:amd64.
(Reading database ... 123638 files and directories currently installed.)
Preparing to unpack .../libhtscodecs2_1.1.1-3_amd64

In [ ]:
# index the reference
!bwa index mycobacterium_tuberculosis_h37rv_2_contigs.fasta

[bwa_index] Pack FASTA... 0.04 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 1.59 seconds elapse.
[bwa_index] Update BWT... 0.04 sec
[bwa_index] Pack forward-only FASTA... 0.03 sec
[bwa_index] Construct SA from BWT and Occ... 0.94 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index mycobacterium_tuberculosis_h37rv_2_contigs.fasta
[main] Real time: 2.756 sec; CPU: 2.643 sec


Here's a brief description of each file:

1.  **`.fasta.amb`**: This file contains information about ambiguous bases in the reference genome. It helps the aligner handle ambiguous nucleotide sequences (like N's) efficiently.

2.  **`.fasta.ann`**: This file is an annotation file that contains information about the sequences, including their names and lengths.

3.  **`.fasta.bwt`**: This is the Burrows-Wheeler Transform index file. It is used to efficiently align reads to the reference genome. The BWT is a reversible transformation that enables efficient data compression and search.

4.  **`.fasta.pac`**: This file contains the packed version of the reference sequence, stored in a compressed binary format to reduce the file size.

5.  **`.fasta.sa`**: This is the suffix array index file. It is used in conjunction with the `.bwt` file to efficiently locate substrings in the reference sequence during alignment.

The following code aligns sequencing reads from the FASTQ file to the reference genome using the `bwa mem` algorithm and outputs the alignments in a SAM file.

What Happens During the Execution ?

1.  **Alignment Process**:

    -   The `bwa mem` algorithm reads the input FASTQ file (`SRR30683486.fastq.gz`) and aligns each read to the reference genome (`mycobacterium_tuberculosis_h37rv_2_supercontigs.fasta`).
    -   It uses the index files created from the reference genome (generated using the `bwa index` command) to find the best alignment positions quickly and efficiently.
2.  **Output in SAM Format**:

    -   The aligned reads, along with detailed alignment information (position, quality, mismatches, etc.), are saved in the `aligned_reads.sam` file.
    -   This SAM file is a standard format used to store alignment information, and it can be converted to BAM format, sorted, and indexed for further analysis.

In [ ]:
# Align the reads with the reference genome using bwa mem
!bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta \
/content/drive/MyDrive/Academic/CMI\ DS\ 23-25/IP/Meeting\ 4:\ 23\ 09\ 2024/SRR30683486.fastq.gz \
> /content/aligned_reads.sam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 68442 sequences (10000014 bp)...
[M::process] read 68372 sequences (10000256 bp)...
[M::mem_process_seqs] Processed 68442 reads in 4.471 CPU sec, 4.262 real sec
[M::process] read 68382 sequences (10000230 bp)...
[M::mem_process_seqs] Processed 68372 reads in 4.641 CPU sec, 4.367 real sec
[M::process] read 68466 sequences (10000125 bp)...
[M::mem_process_seqs] Processed 68382 reads in 5.383 CPU sec, 5.258 real sec
[M::process] read 68416 sequences (10000168 bp)...
[M::mem_process_seqs] Processed 68466 reads in 4.375 CPU sec, 4.310 real sec
[M::process] read 68242 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 68416 reads in 4.468 CPU sec, 4.148 real sec
[M::process] read 68208 sequences (10000175 bp)...
[M::mem_process_seqs] Processed 68242 reads in 5.488 CPU sec, 5.338 real sec
[M::process] read 68166 sequences (10000067 bp)...
[M::mem_process_seqs] Processed 68208 reads in 5.013 CPU sec, 5.025 real sec
[M::pr

## Reading the results of BWA

### What if we do not convert it to BAM?

In [ ]:
!head -n 23 /content/aligned_reads.sam

@SQ	SN:AL123456	LN:4411532
@PG	ID:bwa	PN:bwa	VN:0.7.17-r1188	CL:bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta /content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/SRR30683486.fastq.gz
SRR30683486.1	0	AL123456	454690	60	150M	*	0	0	GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCGCCGGGCCTGATCGCGCTCTTCTCGTCTCGGGCCGACGACCTTTCGGTCGAGTTGAGCGTGCATCCCACCAGCCGGTTCCGCGAACTGATCTGCTCGCGCGCCGT	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:2	MD:Z:7C51A90	AS:i:140	XS:i:0
SRR30683486.1	16	AL123456	455061	60	150M	*	0	0	CGATTCCGGAGTCCCAGCAACGGATCTTCCAGAGCGATGCCGCCGCGCTGGAGGAGGTCATGCGCGTCGGGGGCGGCACGCTGGCCATTGGCTTTGCGGTCGCCAAGGATCTTGCCGCCGGACGGTTGGTGCACGTGACCGGTCCTGGGC	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:1	MD:Z:75C74	AS:i:145	XS:i:0
SRR30683486.

The given line from the SAM file provides detailed alignment information for a single sequencing read (`SRR30683486.1`) against a reference sequence (`AL123456`). Here's a breakdown of each component:

 1\. **Header Lines**

These lines provide meta-information about the reference genome and alignment tools used.

-   **`@SQ SN:AL123456 LN:4411532`**:

    -   **`@SQ`**: Indicates a sequence dictionary entry.
    -   **`SN:AL123456`**: The reference sequence name, in this case, a contig or chromosome labeled `AL123456`.
    -   **`LN:4411532`**: The length of the reference sequence `AL123456` is 4,411,532 base pairs.
-   **`@PG ID:bwa PN:bwa VN:0.7.17-r1188 CL:bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta /content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/SRR30683486.fastq.gz`**:

    -   **`@PG`**: Program information.
    -   **`ID:bwa`**: Program ID is `bwa`.
    -   **`PN:bwa`**: Program name is `bwa`.
    -   **`VN:0.7.17-r1188`**: Version of `bwa` used for alignment.
    -   **`CL:bwa mem ...`**: The command line (`CL`) used to run `bwa mem` with the reference file (`mycobacterium_tuberculosis_h37rv_2_contigs.fasta`) and the input read file (`SRR30683486.fastq.gz`).

 2\. **Alignment Line**

The main part of the SAM file, this line provides details about how a single read aligns to the reference genome.

**Line**:

For the first line only:

`SRR30683486.1	0	AL123456	454690	60	150M	*	0	0	GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCGCCGGGCCTGATCGCGCTCTTCTCGTCTCGGGCCGACGACCTTTCGGTCGAGTTGAGCGTGCATCCCACCAGCCGGTTCCGCGAACTGATCTGCTCGCGCGCCGT	??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????	NM:i:2	MD:Z:7C51A90	AS:i:140	XS:i:0`

| **Field** | **Description** | **Value** |
| --- | --- | --- |
| **QNAME** | Query name (read identifier). | `SRR30683486.1` |
| **FLAG** | Bitwise flag indicating the read's properties. | `0` (indicates the read is mapped to the forward strand) |
| **RNAME** | Reference sequence name where the read is aligned. | `AL123456` (reference sequence or contig name) |
| **POS** | 1-based leftmost position of alignment. | `454690` (the read starts aligning at position 454,690 on the reference sequence `AL123456`) |
| **MAPQ** | Mapping quality score (0-255). | `60` (high confidence in this alignment) |
| **CIGAR** | Describes the alignment in terms of matches, insertions, deletions, etc. | `150M` (150 bases match the reference exactly, no insertions/deletions) |
| **RNEXT** | Reference name of the mate/next read (if paired). | `*` (no mate information, or single-end read) |
| **PNEXT** | Position of the mate/next read (if paired). | `0` (no mate position, or single-end read) |
| **TLEN** | Observed template length (insert size). | `0` (no template length, or single-end read) |
| **SEQ** | Query sequence (the actual read sequence). | `GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCGCCGGGCCTGATCGCGCTCTTCTCGTCTCGGGCCGACGACCTTTCGGTCGAGTTGAGCGTGCATCCCACCAGCCGGTTCCGCGAACTGATCTGCTCGCGCGCCGT` |
| **QUAL** | Phred-scaled base quality scores. | `??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????` (indicates quality is missing or not available) |
| **Optional Tags** | Additional information about the alignment. | `NM:i:2`, `MD:Z:7C51A90`, `AS:i:140`, `XS:i:0` |

 Detailed Explanation of Key Fields

1.  **QNAME (Query Name)**:

    -   **`SRR30683486.1`**: The unique identifier for the read. This allows you to trace this specific read in your dataset.
2.  **FLAG**:

    -   **`0`**: Indicates the read is mapped to the forward strand with no special properties.
        -   `0` means the read is mapped and is a primary alignment.
3.  **RNAME (Reference Name)**:

    -   **`AL123456`**: The reference sequence or contig to which the read is aligned.
4.  **POS (Position)**:

    -   **`454690`**: The starting position of the read alignment on the reference sequence `AL123456`. This means the first base of the read aligns to position 454,690 of the reference.
5.  **MAPQ (Mapping Quality)**:

    -   **`60`**: Indicates a high confidence in the read alignment. A MAPQ of 60 is the maximum score, signifying the alignment is unique with very high certainty.
6.  **CIGAR (Compact Idiosyncratic Gapped Alignment Report)**:

    -   **`150M`**: This indicates that all 150 bases in the read match exactly to the reference genome without any insertions, deletions, or clipping.
        -   `150M` means 150 bases are aligned with either matches or mismatches but no gaps or clipping.
7.  **SEQ (Sequence)**:

    -   The sequence of the read that has been aligned to the reference.
    -   **`GTTGCTGAGCATCGCC...`**: This is the nucleotide sequence of the read.
8.  **QUAL (Quality Scores)**:

    -   The quality scores for each base in the read. Typically, these are ASCII-encoded Phred scores representing the confidence in each base call.
    -   **`????????????...`**: In this case, it appears as `?`, which may indicate missing or placeholder quality scores, suggesting the quality information is not present.
9.  **Optional Tags**:

    -   Provide additional details about the alignment:

    -   **`NM:i:2`**: Number of mismatches between the read and the reference sequence. Here, there are 2 mismatches.

    -   **`MD:Z:7C51A90`**: Mismatch descriptor tag:

        -   `7C51A90` means:
            -   `7`: First 7 bases match exactly.
            -   `C`: Mismatch at the 8th position (reference has `C`).
            -   `51`: Next 51 bases match exactly.
            -   `A`: Mismatch at the 60th position (reference has `A`).
            -   `90`: Last 90 bases match exactly.
    -   **`AS:i:140`**: Alignment score, a measure of how well the read aligns to the reference. Higher scores indicate better alignments.

    -   **`XS:i:0`**: Suboptimal alignment score. If there were multiple potential alignments for this read, this would show the score of the next best alignment. Here, `0` suggests no suboptimal alignment.

 Summary

This alignment line provides a detailed view of how the read `SRR30683486.1` aligns to the reference sequence `AL123456`:

-   The read aligns starting at position `454690` on the reference with a high-quality score of `60`.
-   The CIGAR string `150M` indicates that all 150 bases of the read match exactly to the reference sequence.
-   There are 2 mismatches between the read and the reference, as indicated by the `NM:i:2` tag.
-   The `MD:Z:7C51A90` tag specifies the exact positions of these mismatches.

This information is crucial for understanding the accuracy and quality of the alignment, which is necessary for downstream analyses like variant calling or gene expression studies. If you need more details or specific interpretations, feel free to ask!

In [ ]:
# !cat /content/aligned_reads.sam

Total number of lines

In [ ]:
!wc -l /content/aligned_reads.sam

574362 /content/aligned_reads.sam


In [ ]:
574362*150/(2*4.4*10**6)

9.790261363636363

To find the total number of discrepancies (mismatches, insertions, and deletions) between the reads in your SAM file and the reference genome, you can use the samtools tool along with some additional processing.

In [ ]:
!apt-get install -y samtools
!samtools view -Sb /content/aligned_reads.sam > /content/aligned_reads.bam
#!samtools view -Sb /content/aligned_reads.sam > /content/aligned_reads.bam
!samtools view /content/aligned_reads.bam | awk '{for(i=12;i<=NF;i++) if($i ~ /^NM:i:/) {split($i,a,":"); total += a[3]}} END {print "Total number of discrepancies:", total}'

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
samtools is already the newest version (1.13-4).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.
Total number of discrepancies: 249797


In [ ]:
import pandas as pd

# Read the first 20 lines of the SAM file, skipping lines with parsing errors
sam_file = '/content/aligned_reads.sam'
sam_data = pd.read_csv(
    sam_file,
    sep='\t',
    comment='@',
    header=None,
    names=None,
    on_bad_lines='skip',  # Skip lines with mismatched columns  # Read the first 20 lines
)

# Display the DataFrame
sam_data

0   1         2        3   4        5  6   7   8   \
0            SRR30683486.1   0  AL123456   454690  60     150M  *   0   0   
1            SRR30683486.1  16  AL123456   455061  60     150M  *   0   0   
2            SRR30683486.2  16  AL123456  1425788  60     151M  *   0   0   
3            SRR30683486.2   0  AL123456  1425435  60     148M  *   0   0   
4            SRR30683486.3  16  AL123456  3102028  60     151M  *   0   0   
...                    ...  ..       ...      ...  ..      ... ..  ..  ..   
565483  SRR30683486.286531  16  AL123456  4173889  60     146M  *   0   0   
565484  SRR30683486.286532  16  AL123456  2798987  60     150M  *   0   0   
565485  SRR30683486.286532   0  AL123456  2798544  60     114M  *   0   0   
565486  SRR30683486.286533  16  AL123456  3633106  60     151M  *   0   0   
565487  SRR30683486.286533   0  AL123456  3632971  60  116M35S  *   0   0   

                                                       9   \
0       GTTGCTGAGCATCGCCGCCTCCAGCGCCTTCGCCGAACACGCCGCG...   
1       CGATTCCGGAGTCCCAGCAACGGATCTTCCAGAGCGATGCCGCCGC...   
2       CATCGACGGCCGGCAGGTTGGCGCGCAGCCAGCCACCGGCCAGCCC...   
3       CCTCTACCGTGGCACGTGAAAGCCGACCAGCGCGGCACACCCCGGT...   
4       ATCGGCCGGCTGGTCGTCCACGAGAATCAGCCGCCCGGCCTTCCGC...   
...                                                   ...   
565483  ATCTGACGCTGCGAGAGCTGGCCAAATCAGCTCATGTCAAGCGATA...   
565484  CACGCCACAACGGCTGCAGACATGATGCCAACAGCAACGCGTGGCC...   
565485  AAGGAAGGCATCACCAATCTCGTCGGCGAGGCTTCGTGCCTCCTCG...   
565486  AACGTGACGATCACGAGCGCACCCCAGTCGAGCTGCAGCGATCCGA...   
565487  GGCGAGTAGTTGGCGGCTCTGATGTAGAAGGCACCGACGACCACCA...   

                                                       10      11  \
0       ??????????????????????????????????????????????...  NM:i:2   
1       ??????????????????????????????????????????????...  NM:i:1   
2       ??????????????????????????????????????????????...  NM:i:0   
3       ??????????????????????????????????????????????...  NM:i:0   
4       ??????????????????????????????????????????????...  NM:i:0   
...                                                   ...     ...   
565483  ??????????????????????????????????????????????...  NM:i:0   
565484  ??????????????????????????????????????????????...  NM:i:0   
565485  ??????????????????????????????????????????????...  NM:i:1   
565486  ??????????????????????????????????????????????...  NM:i:1   
565487  ??????????????????????????????????????????????...  NM:i:3   

                     12        13       14  
0          MD:Z:7C51A90  AS:i:140   XS:i:0  
1            MD:Z:75C74  AS:i:145   XS:i:0  
2              MD:Z:151  AS:i:151   XS:i:0  
3              MD:Z:148  AS:i:148   XS:i:0  
4              MD:Z:151  AS:i:151   XS:i:0  
...                 ...       ...      ...  
565483         MD:Z:146  AS:i:146  XS:i:20  
565484         MD:Z:150  AS:i:150   XS:i:0  
565485       MD:Z:6T107  AS:i:109   XS:i:0  
565486      MD:Z:119A31  AS:i:146   XS:i:0  
565487  MD:Z:17C78G6A12  AS:i:107   XS:i:0  

[565488 rows x 15 columns]

In [ ]:
565488-286533*2

-7578

In [ ]:
sum = 0
for i in sam_data[11]:
  sum+= int(i[-1])
print(sum) #== #249797

209695


# Reading and using Picard

### 1\. **Sorting and Merging BAM/SAM Files**

-   **Sorting**: Think of sorting as arranging your data in a particular order. For BAM/SAM files, this usually means ordering the reads based on their position in the genome. This makes it easier to find and analyze specific regions in your data.
-   **Merging**: If you have multiple files containing parts of your data, merging them combines all these parts into a single, complete file. This is useful when you have data from different experiments or runs and want to analyze them together.

### 2\. **Marking Duplicates**

-   During sequencing, sometimes the same piece of DNA gets copied multiple times. Picard can find these duplicate copies and mark them. By marking duplicates, you can ignore them in your analysis, so your results aren't skewed by counting the same piece of DNA more than once.

### 3\. **Collecting Quality Metrics**

-   Picard can check the quality of your sequencing data by calculating various statistics. For example, it can tell you the average size of DNA fragments, how well your reads align to the reference genome, and if there are any unusual patterns. This helps ensure your data is good enough for further analysis.

### 4\. **File Conversion**

-   Picard can change the format of your data files. For example, it can switch between BAM (a compressed, efficient format) and SAM (a plain-text version) formats. This is helpful when different tools require different formats.

### 5\. **Manipulating Headers**

-   The header of a BAM/SAM file contains important information about the data, like the sample name and sequencing details. Picard allows you to change this information, such as adding new details or correcting mistakes, so that your files have accurate and complete metadata.

Picard is a Java-based tool, so you need to install Java first.

In [ ]:
!apt-get install -y openjdk-11-jdk

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java libatk-wrapper-java-jni libfontenc1
  libice-dev libsm-dev libxkbfile1 libxt-dev libxtst6 libxxf86dga1 openjdk-11-jre x11-utils
Suggested packages:
  libice-doc libsm-doc libxt-doc openjdk-11-demo openjdk-11-source visualvm mesa-utils
The following NEW packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java libatk-wrapper-java-jni libfontenc1
  libice-dev libsm-dev libxkbfile1 libxt-dev libxtst6 libxxf86dga1 openjdk-11-jdk openjdk-11-jre
  x11-utils
0 upgraded, 14 newly installed, 0 to remove and 49 not upgraded.
Need to get 5,517 kB of archives.
After this operation, 15.8 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 fonts-dejavu-core all 2.37-2build1 [1,041 kB]
Get:2 http://archive.ubuntu.com/ubun

You can download the Picard JAR file from the official Broad Institute website.

In [ ]:
!wget -O picard.jar https://github.com/broadinstitute/picard/releases/download/2.27.4/picard.jar

--2024-10-11 14:06:40--  https://github.com/broadinstitute/picard/releases/download/2.27.4/picard.jar
Resolving github.com (github.com)... 20.27.177.113
Connecting to github.com (github.com)|20.27.177.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/18225913/839cd9dd-e7dc-4c29-ab0d-6ab4bdcded4e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241011%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241011T140640Z&X-Amz-Expires=300&X-Amz-Signature=bebc94c3fbbab5ecd7345eb5fbd028571b0b48cdd2d21c13d2cdbf81721b359d&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dpicard.jar&response-content-type=application%2Foctet-stream [following]
--2024-10-11 14:06:40--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/18225913/839cd9dd-e7dc-4c29-ab0d-6ab4bdcded4e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential

This command uses Picard's `SortSam` tool to:

1.  **Input**: Read the BAM file (`aligned_reads.bam`).
2.  **Output**: Write a new BAM file (`sorted_reads.bam`).
3.  **Action**: Sort the reads in the BAM file by their genomic coordinates (`SORT_ORDER=coordinate`), making it easier to access and analyze specific regions of the genome.

In [ ]:
!java -jar picard.jar SortSam \
    INPUT=/content/aligned_reads.bam \
    OUTPUT=/content/sorted_reads.bam \
    SORT_ORDER=coordinate

INFO	2024-10-11 14:06:43	SortSam	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    SortSam -INPUT /content/aligned_reads.bam -OUTPUT /content/sorted_reads.bam -SORT_ORDER coordinate
**********


14:06:44.204 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Fri Oct 11 14:06:44 UTC 2024] SortSam INPUT=/content/aligned_reads.bam OUTPUT=/content/sorted_reads.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false
[Fri Oct 11 14

Key Points in the Log:

1.  **Command Line Syntax Change Notification**:

    -   Picard is notifying you that the command line syntax is changing. They are moving to a new syntax format. The example given in the log shows how the new command syntax would look:

        bash

        `SortSam -INPUT /content/aligned_reads.bam -OUTPUT /content/sorted_reads.bam -SORT_ORDER coordinate`

2.  **Library Loading**:

    -   `INFO NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!...`
    -   This indicates that Picard is loading native libraries required for compression and decompression operations, which are used to handle the BAM file format efficiently.
3.  **Execution Information**:

    -   The line `[Mon Sep 23 05:25:43 UTC 2024] SortSam INPUT=/content/aligned_reads.bam OUTPUT=/content/sorted_reads.bam ...`
    -   This shows the command being executed with various parameters:
        -   `INPUT`: The input BAM file.
        -   `OUTPUT`: The output file to which the sorted reads are written.
        -   `SORT_ORDER`: The order in which the reads are sorted (in this case, `coordinate`).
        -   `VERBOSITY=INFO`: The level of logging information.
        -   `VALIDATION_STRINGENCY=STRICT`: How strictly Picard checks the input file for errors.
        -   Other parameters control compression, indexing, memory usage, and more.
4.  **Non-Increasing Record Positions**:

    -   `INFO 2024-09-23 05:25:43 SortSam Seen many non-increasing record positions. Printing Read-names as well.`
    -   This message suggests that the input BAM file had records (reads) that were not in the expected order (which is why you're sorting them). This is typical when sorting BAM files.
5.  **Processing Complete**:

    -   `INFO 2024-09-23 05:25:51 SortSam Finished reading inputs, merging and writing to output now.`
    -   This indicates that Picard has finished reading the input BAM file, merged the reads in the correct order, and is now writing the sorted reads to the output BAM file.
6.  **Execution Finished**:

    -   `[Mon Sep 23 05:25:56 UTC 2024] picard.sam.SortSam done. Elapsed time: 0.22 minutes.`
    -   This means the entire sorting process is complete, and it took around 0.22 minutes (about 13 seconds) to finish the task.
7.  **Memory Usage**:

    -   `Runtime.totalMemory()=124780544`
    -   This is the amount of memory (in bytes) allocated to the Java Virtual Machine (JVM) during the execution.

### Summary:

The log indicates that the `SortSam` command successfully sorted the input BAM file, despite encountering many non-increasing record positions (which is normal for unsorted BAM files). The new syntax notification is just a heads-up for future command usage. The operation was completed in around 13 seconds, and everything seems to have worked as expected.

In [ ]:
!java -jar picard.jar MarkDuplicates \
    INPUT=/content/sorted_reads.bam \
    OUTPUT=/content/dedup_reads.bam \
    METRICS_FILE=/content/duplication_metrics.txt

INFO	2024-10-11 14:06:57	MarkDuplicates	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    MarkDuplicates -INPUT /content/sorted_reads.bam -OUTPUT /content/dedup_reads.bam -METRICS_FILE /content/duplication_metrics.txt
**********


14:06:58.426 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Fri Oct 11 14:06:58 UTC 2024] MarkDuplicates INPUT=[/content/sorted_reads.bam] OUTPUT=/content/dedup_reads.bam METRICS_FILE=/content/duplication_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATE

The log you've shared is the detailed output of running the `MarkDuplicates` command using Picard. Here's a step-by-step explanation of the information provided in the log:

Key Points in the Log:

1.  **Command Line Syntax Change Notification**:

    -   Similar to `SortSam`, Picard is informing you that the command line syntax is changing. The log provides the new format of the `MarkDuplicates` command:

        bash

        Copy code

        `MarkDuplicates -INPUT /content/sorted_reads.bam -OUTPUT /content/dedup_reads.bam -METRICS_FILE /content/duplication_metrics.txt`

2.  **Library Loading**:

    -   `INFO NativeLibraryLoader - Loading libgkl_compression.so...`
    -   Picard is loading native libraries for compression, which are required to handle BAM files efficiently.
3.  **Command Execution Details**:

    -   The command being executed is:

        javascript

        Copy code

        `MarkDuplicates INPUT=[/content/sorted_reads.bam] OUTPUT=/content/dedup_reads.bam METRICS_FILE=/content/duplication_metrics.txt ...`

    -   Key parameters used:
        -   `INPUT`: Specifies the input sorted BAM file.
        -   `OUTPUT`: Specifies the output BAM file with duplicates marked.
        -   `METRICS_FILE`: Specifies the file where duplication metrics (e.g., number of duplicates) are written.
        -   Various other parameters control aspects like memory usage, duplicate scoring, and how read names are parsed.
4.  **Start of Processing**:

    -   `INFO MarkDuplicates Start of doWork freeMemory: 24723776; totalMemory: 31457280; maxMemory: 3403677696`
    -   Shows initial memory allocation for the Java process before starting the duplicate marking operation.
5.  **Reading and Constructing Read End Information**:

    -   `INFO MarkDuplicates Reading input file and constructing read end information.`
    -   Picard is reading the input BAM file to gather information about the reads' ends (the start and end positions of each read) to identify duplicates.
6.  **Non-Integer Read Name Warning**:

    -   `WARNING AbstractOpticalDuplicateFinderCommandLineProgram ... Read name: SRR30683486.118524. Cause: String 'SRR30683486.118524' did not start with a parsable number.`
    -   This warning indicates that a part of a read name, which Picard expected to contain an integer, did not. This is usually not a critical issue but indicates that the read naming convention is not as expected by Picard.
7.  **Number of Records Read**:

    -   `INFO MarkDuplicates Read 568784 records. 0 pairs never matched.`
    -   Picard read 568,784 records from the BAM file and did not encounter any unmatched read pairs, which means all paired-end reads were properly aligned.
8.  **Memory Usage Information**:

    -   Several memory usage statistics are logged throughout the process, indicating how much memory Picard is using as it processes the file.
9.  **Duplicate Detection**:

    -   `INFO MarkDuplicates Traversing read pair information and detecting duplicates.`
    -   Picard is now checking each read pair for duplicates.
10. **Duplicates Marked**:

    -   `INFO MarkDuplicates Marking 102504 records as duplicates.`
    -   102,504 records were identified and marked as duplicates.
11. **Optical Duplicate Detection**:

    -   `INFO MarkDuplicates Found 0 optical duplicate clusters.`
    -   No optical duplicates (duplicates caused by the imaging system of the sequencer) were found. This is good, as it suggests no imaging artifacts.
12. **Writing Complete**:

    -   `INFO MarkDuplicates Writing complete. Closing input iterator.`
    -   Picard has finished writing the output BAM file with duplicates marked and is closing the input and output files.
13. **Duplicate Index Cleanup**:

    -   `INFO MarkDuplicates Duplicate Index cleanup.`
    -   Picard is cleaning up any temporary indices it created during the processing.
14. **Final Memory Statistics and Output Closure**:

    -   Shows the memory usage statistics before and after closing the output file.
15. **Execution Complete**:

    -   `[Mon Sep 23 05:49:11 UTC 2024] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.28 minutes.`
    -   The `MarkDuplicates` command has finished execution in approximately 0.28 minutes (about 17 seconds).
16. **Final Memory Usage**:

    -   `Runtime.totalMemory()=28311552`
    -   This shows the final memory allocated to the JVM after completing the command.

Summary:

The `MarkDuplicates` command successfully identified and marked duplicates in the BAM file. It read 568,784 records and marked 102,504 of them as duplicates. The process took around 17 seconds to complete, and no optical duplicates were found. The log also includes detailed information on memory usage and processing steps, indicating that everything completed successfully without critical errors.


In [ ]:
#convert deduplicated BAM file to SAM format
!java -jar picard.jar SamFormatConverter \
    INPUT=/content/dedup_reads.bam \
    OUTPUT=/content/dedup_reads.sam

INFO	2024-10-11 14:07:13	SamFormatConverter	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    SamFormatConverter -INPUT /content/aligned_reads.bam -OUTPUT /content/aligned_reads.sam
**********


14:07:14.738 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Fri Oct 11 14:07:14 UTC 2024] SamFormatConverter INPUT=/content/aligned_reads.bam OUTPUT=/content/aligned_reads.sam    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false
[Fri Oct 11 14:07:14 UTC

 Command Breakdown:

`!java -jar picard.jar AddOrReplaceReadGroups\
    INPUT=/content/aligned_reads.bam\
    OUTPUT=/content/rg_aligned_reads.bam\
    RGID=1\
    RGLB=lib1\
    RGPL=illumina\
    RGPU=unit1\
    RGSM=sample`

-   `!java -jar picard.jar AddOrReplaceReadGroups \`:

    -   This is the main command to run Picard's `AddOrReplaceReadGroups` tool using the Java command-line interface.
-   `INPUT=/content/aligned_reads.bam \`:

    -   Specifies the path to the input BAM file (`/content/aligned_reads.bam`) where the read group information will be added or replaced.
-   `OUTPUT=/content/rg_aligned_reads.bam \`:

    -   Specifies the path to the output BAM file (`/content/rg_aligned_reads.bam`) with the updated read group information.
-   `RGID=1 \`:

    -   `RGID` stands for "Read Group ID." It uniquely identifies a read group within the BAM file. Here, it is set to `1`.
-   `RGLB=lib1 \`:

    -   `RGLB` stands for "Read Group Library." It specifies the library preparation kit or process used. Here, it is set to `lib1`.
-   `RGPL=illumina \`:

    -   `RGPL` stands for "Read Group Platform." It specifies the sequencing platform used. In this case, `illumina` is specified, which is a commonly used sequencing technology.
-   `RGPU=unit1 \`:

    -   `RGPU` stands for "Read Group Platform Unit." It typically refers to the run or flow cell barcode used for sequencing. Here, it is set to `unit1`.
-   `RGSM=sample`:

    -   `RGSM` stands for "Read Group Sample." It specifies the name of the sample being sequenced. Here, it is set to `sample`.

 Why Is This Important?

1.  **Metadata Annotation**: Adding read group information helps in keeping track of different sequencing runs, libraries, and samples, which is crucial for organizing and analyzing complex datasets.

2.  **Downstream Analysis**: Many tools, like GATK (Genome Analysis Toolkit) for variant calling, require read group information to differentiate between data from different samples or sequencing runs within a single BAM file.

3.  **Quality Control**: By including details about the platform and library, researchers can perform quality checks and identify issues specific to a particular run or library preparation.

 Summary:

This command adds or updates metadata in your BAM file to include specific details about the sequencing process and sample. This information is essential for proper data management and accurate downstream analysis.

In [ ]:
!head /content/dedup_reads.sam

@HD	VN:1.6	SO:coordinate
@SQ	SN:AL123456	LN:4411532
@PG	ID:bwa	PN:bwa	VN:0.7.17-r1188	CL:bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta /content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/SRR30683486.fastq.gz
@PG	ID:samtools	PN:samtools	PP:bwa	VN:1.13	CL:samtools view -Sb /content/aligned_reads.sam
@PG	ID:MarkDuplicates	VN:2.27.4-SNAPSHOT	CL:MarkDuplicates INPUT=[/content/sorted_reads.bam] OUTPUT=/content/dedup_reads.bam METRICS_FILE=/content/duplication_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_QUALITY_SUM_STRATEGY=false USE_END_IN_UNPAIRED_READS=false USE_UNPAIRED_CLIPPED_END=false UNPAIRED_END_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true

In [ ]:
!head -n 50 /content/dedup_reads.sam

@HD	VN:1.6	SO:coordinate
@SQ	SN:AL123456	LN:4411532
@PG	ID:bwa	PN:bwa	VN:0.7.17-r1188	CL:bwa mem /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta /content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/SRR30683486.fastq.gz
@PG	ID:samtools	PN:samtools	PP:bwa	VN:1.13	CL:samtools view -Sb /content/aligned_reads.sam
@PG	ID:MarkDuplicates	VN:2.27.4-SNAPSHOT	CL:MarkDuplicates INPUT=[/content/sorted_reads.bam] OUTPUT=/content/dedup_reads.bam METRICS_FILE=/content/duplication_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_QUALITY_SUM_STRATEGY=false USE_END_IN_UNPAIRED_READS=false USE_UNPAIRED_CLIPPED_END=false UNPAIRED_END_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true

In [ ]:
# !cat /content/dedup_reads.sam

## Base recalibration

**Why Perform Base Quality Score Recalibration (BQSR)?**

1.  **Correcting Systematic Bias**: During sequencing, errors can occur, and sometimes the quality scores (which tell us how confident the machine is in calling each base, like A, T, C, or G) might not be accurate. These errors can happen in a consistent, or systematic, way, such as due to PCR amplification. BQSR helps fix these issues, making the quality scores more reliable.

2.  **Improving Accuracy**: After recalibrating the base quality scores, we get a better and more accurate measure of how confident we are in each base call. This helps improve the accuracy of future analyses, such as finding genetic variants (differences in DNA) between individuals.

3.  **Best Practice in Variant Discovery**: Recalibrating quality scores is considered a standard, or "best practice," before doing variant calling (finding differences in DNA) in workflows like GATK's pipeline. It ensures that the results are more reliable and accurate.

In simpler terms, BQSR corrects potential errors in the quality scores assigned to each base during sequencing, which helps make downstream analysis, like finding genetic variations, more accurate and trustworthy.

Install GATK, Check GATK Installation:

In [ ]:
!wget -c https://github.com/broadinstitute/gatk/releases/download/4.3.0.0/gatk-4.3.0.0.zip

!unzip -o gatk-4.3.0.0.zip

!gatk-4.3.0.0/gatk --version

--2024-10-11 15:00:27--  https://github.com/broadinstitute/gatk/releases/download/4.3.0.0/gatk-4.3.0.0.zip
Resolving github.com (github.com)... 20.27.177.113
Connecting to github.com (github.com)|20.27.177.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/27452807/2f011c35-f7ee-4b53-8236-75558badeb82?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241011%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241011T150027Z&X-Amz-Expires=300&X-Amz-Signature=3f0b434a06c3c7f2febc9016733be7537846a23f18777f31926f31b77f5bc046&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dgatk-4.3.0.0.zip&response-content-type=application%2Foctet-stream [following]
--2024-10-11 15:00:28--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/27452807/2f011c35-f7ee-4b53-8236-75558badeb82?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz

the reference genome name `M.tuberculosis_H37Rv` in the VCF file is being replaced with `AL123456` to match the reference genome name used in the BAM/SAM file. Ensuring consistent naming between files is crucial when performing downstream analysis like base quality score recalibration or variant calling, where the tools need the reference genome name in all files to match exactly.

In [ ]:
# Replace 'M.tuberculosis_H37Rv' with 'AL123456' in the VCF file
!sed -i 's/M.tuberculosis_H37Rv/AL123456/g' "/content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/MTB_Base_Calibration_List.vcf"

Indexing the VCF file speeds up processes that require access to specific regions of the VCF file, such as during **Base Quality Score Recalibration (BQSR)**. The index file allows GATK tools to quickly find the information they need without scanning the entire VCF file.

In [ ]:
!java -jar /content/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar IndexFeatureFile \
   -I "/content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/MTB_Base_Calibration_List.vcf"

14:09:40.004 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
14:09:40.250 INFO  IndexFeatureFile - ------------------------------------------------------------
14:09:40.251 INFO  IndexFeatureFile - The Genome Analysis Toolkit (GATK) v4.3.0.0
14:09:40.251 INFO  IndexFeatureFile - For support and documentation go to https://software.broadinstitute.org/gatk/
14:09:40.252 INFO  IndexFeatureFile - Executing as root@594aca40fa34 on Linux v6.1.85+ amd64
14:09:40.252 INFO  IndexFeatureFile - Java runtime: OpenJDK 64-Bit Server VM v11.0.24+8-post-Ubuntu-1ubuntu322.04
14:09:40.252 INFO  IndexFeatureFile - Start Date/Time: October 11, 2024 at 2:09:39 PM UTC
14:09:40.252 INFO  IndexFeatureFile - ------------------------------------------------------------
14:09:40.253 INFO  IndexFeatureFile - ------------------------------------------------------------
14:09:40.254 INFO  IndexFe

The log confirms that the VCF file has been successfully indexed, and the index file (`.idx`) is now ready to be used for faster access during downstream processes like **Base Quality Score Recalibration (BQSR)**. Everything looks good!

After running the following commands, you'll have a new BAM file (`rg_dedup_reads.bam`) that contains the read group information. This file is now ready for further analyses, such as base quality score recalibration or variant calling.

-   The **first command** downloads Picard.
-   The **second command** adds read group information to the deduplicated BAM file, which is important for downstream analysis and variant calling.

In [ ]:
!wget -c https://github.com/broadinstitute/picard/releases/download/2.27.4/picard.jar

!java -jar picard.jar AddOrReplaceReadGroups \
   I=/content/dedup_reads.bam \
   O=/content/rg_dedup_reads.bam \
   RGID=1 \
   RGLB=lib1 \
   RGPL=illumina \
   RGPU=unit1 \
   RGSM=sample

--2024-10-11 14:09:41--  https://github.com/broadinstitute/picard/releases/download/2.27.4/picard.jar
Resolving github.com (github.com)... 20.27.177.113
Connecting to github.com (github.com)|20.27.177.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/18225913/839cd9dd-e7dc-4c29-ab0d-6ab4bdcded4e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241011%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241011T140941Z&X-Amz-Expires=300&X-Amz-Signature=b28084cd64dee04db718ba23338237510b87b36d7b6fedde5bab6021946aa349&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dpicard.jar&response-content-type=application%2Foctet-stream [following]
--2024-10-11 14:09:41--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/18225913/839cd9dd-e7dc-4c29-ab0d-6ab4bdcded4e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential

Index the rg_dedup_reads.bam bam file

In [ ]:
!samtools index /content/rg_dedup_reads.bam

This command is using the BaseRecalibrator tool from the Genome Analysis Toolkit (GATK) to perform base quality score recalibration (BQSR) on your BAM file. Here's a detailed breakdown of what each part of the command does:

Now, You need to create the index file for your reference genome. GATK requires both a FASTA index (`.fai`) and a dictionary (`.dict`) for the reference genome in order to perform recalibration.You can do this with `samtools` using the following command:


In [ ]:
!samtools faidx /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta

!java -jar picard.jar CreateSequenceDictionary \
    R=/content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta \
    O=/content/mycobacterium_tuberculosis_h37rv_2_contigs.dict

INFO	2024-10-11 16:54:49	CreateSequenceDictionary	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    CreateSequenceDictionary -R /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta -O /content/mycobacterium_tuberculosis_h37rv_2_contigs.dict
**********


16:54:50.525 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/picard.jar!/com/intel/gkl/native/libgkl_compression.so
[Fri Oct 11 16:54:50 UTC 2024] CreateSequenceDictionary OUTPUT=/content/mycobacterium_tuberculosis_h37rv_2_contigs.dict REFERENCE=/content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMP

The following command you're running is using **GATK's BaseRecalibrator** to generate a **recalibration table** for your BAM file. This process adjusts the base quality scores in your sequencing data, correcting for biases introduced during the sequencing process. Here's a detailed breakdown of what the command does

In [ ]:
!java -Dsamjdk.use_async_io_read_samtools=false \
    -Dsamjdk.use_async_io_write_samtools=true \
    -Dsamjdk.use_async_io_write_tribble=false \
    -Dsamjdk.compression_level=2 \
    -jar /content/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar BaseRecalibrator \
   -I /content/rg_dedup_reads.bam \
   -R /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta \
   --known-sites "/content/drive/MyDrive/Academic/CMI DS 23-25/IP/Meeting 4: 23 09 2024/MTB_Base_Calibration_List.vcf" \
   -O /content/recal_data.table

16:55:12.981 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
16:55:13.239 INFO  BaseRecalibrator - ------------------------------------------------------------
16:55:13.239 INFO  BaseRecalibrator - The Genome Analysis Toolkit (GATK) v4.3.0.0
16:55:13.239 INFO  BaseRecalibrator - For support and documentation go to https://software.broadinstitute.org/gatk/
16:55:13.240 INFO  BaseRecalibrator - Executing as root@594aca40fa34 on Linux v6.1.85+ amd64
16:55:13.240 INFO  BaseRecalibrator - Java runtime: OpenJDK 64-Bit Server VM v11.0.24+8-post-Ubuntu-1ubuntu322.04
16:55:13.241 INFO  BaseRecalibrator - Start Date/Time: October 11, 2024 at 4:55:12 PM UTC
16:55:13.241 INFO  BaseRecalibrator - ------------------------------------------------------------
16:55:13.241 INFO  BaseRecalibrator - ------------------------------------------------------------
16:55:13.242 INFO  BaseRec

Your command executed successfully! Here's what happened during the run:

-   **Initialization**: GATK's `BaseRecalibrator` started by loading the required reference genome and VCF file containing known variants.

-   **Covariates**: It used the following covariates to model the base quality score recalibration:

    -   **Read Group**
    -   **Quality Score**
    -   **Context**
    -   **Cycle**
-   **Progress**: The tool processed 460,125 reads, filtering out:

    -   **12982 reads** due to a quality score of zero.
    -   **101253 reads** because they were marked as duplicates.
    -   **0 reads** were filtered by other criteria.
-   **Completion**: The recalibration was successfully applied to 460,125 reads, and a recalibration report (likely stored in `/content/recal_data.table`) was generated.

This recalibration table can now be used to **apply base quality recalibration** to your BAM file in the next step.

Now that the recalibration table has been generated, you need to apply the recalibration to your BAM file using the ApplyBQSR tool. Here's the command:

In [ ]:
!java -Dsamjdk.use_async_io_read_samtools=false \
    -Dsamjdk.use_async_io_write_samtools=true \
    -Dsamjdk.use_async_io_write_tribble=false \
    -Dsamjdk.compression_level=2 \
    -jar /content/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar ApplyBQSR \
   -R /content/mycobacterium_tuberculosis_h37rv_2_contigs.fasta \
   -I /content/rg_dedup_reads.bam \
   --bqsr-recal-file /content/recal_data.table \
   -O /content/recalibrated_reads.bam

16:59:18.363 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/content/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
16:59:18.589 INFO  ApplyBQSR - ------------------------------------------------------------
16:59:18.589 INFO  ApplyBQSR - The Genome Analysis Toolkit (GATK) v4.3.0.0
16:59:18.590 INFO  ApplyBQSR - For support and documentation go to https://software.broadinstitute.org/gatk/
16:59:18.590 INFO  ApplyBQSR - Executing as root@594aca40fa34 on Linux v6.1.85+ amd64
16:59:18.590 INFO  ApplyBQSR - Java runtime: OpenJDK 64-Bit Server VM v11.0.24+8-post-Ubuntu-1ubuntu322.04
16:59:18.590 INFO  ApplyBQSR - Start Date/Time: October 11, 2024 at 4:59:18 PM UTC
16:59:18.591 INFO  ApplyBQSR - ------------------------------------------------------------
16:59:18.591 INFO  ApplyBQSR - ------------------------------------------------------------
16:59:18.592 INFO  ApplyBQSR - HTSJDK Version: 3.0.1
16:59:18.592 INFO  ApplyBQSR 

The output indicates that the **ApplyBQSR** process ran successfully. Here's a breakdown:

-   **Initialization**: The tool loaded required libraries and initialized successfully, using the specified BAM file (`rg_dedup_reads.bam`), reference genome (`mycobacterium_tuberculosis_h37rv_2_contigs.fasta`), and recalibration table (`recal_data.table`).

-   **Progress**:

    -   **Processed Reads**: A total of **574,360 reads** were processed in 0.43 minutes.
    -   **No reads were filtered** by the `WellformedReadFilter`, meaning all reads passed the basic well-formedness checks.
    -   **Traversal rate**: The tool processed reads at a speed of about **1.36 million reads per minute**, which is efficient.
-   **Completion**: The recalibration was applied successfully, and the process was completed without any errors or warnings (except for a minor "Zero Bytes Written" warning which is not critical).

You now have a recalibrated BAM file (`recalibrated_reads.bam`) ready for further downstream analyses such as variant calling.

Would you like to proceed with the next steps?

4o

After applying the base recalibration using ApplyBQSR, your reads in the BAM file (recalibrated_reads.bam) are considered "analysis-ready." These reads have corrected base quality scores that better reflect the true accuracy of the sequencing data, making them suitable for downstream analysis such as variant calling.

In [ ]:
!samtools flagstat /content/recalibrated_reads.bam

574360 + 0 in total (QC-passed reads + QC-failed reads)
573066 + 0 primary
0 + 0 secondary
1294 + 0 supplementary
102504 + 0 duplicates
102504 + 0 primary duplicates
568784 + 0 mapped (99.03% : N/A)
567490 + 0 primary mapped (99.03% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


### Summary of Output and Interpretation:

-   **Total Reads**: `574,360`

    -   All reads passed QC, indicating no major issues with the sequencing quality.
-   **Primary Reads**: `573,066`

    -   These are the primary alignments (main alignments for each read).
-   **Supplementary Reads**: `1,294`

    -   These are reads that are part of chimeric alignments or split reads, typical for structural variants.
-   **Duplicate Reads**: `102,504`

    -   These represent duplicates, which could have arisen due to PCR amplification during library preparation. These duplicates should be excluded from variant calling as they might introduce bias.
-   **Mapped Reads**: `568,784` (99.03%)

    -   99.03% of reads are mapped to the reference genome, which is a good mapping rate and indicates high-quality alignment.
-   **Paired-End Information**: `0`

    -   Since the reads are single-end, all fields related to paired-end reads are zero, which is expected.
-   **Singletons and Mate Pair Info**: `0`

    -   Again, this is expected as there are no paired-end reads.

### Interpretation:

The output shows that most of your reads have mapped successfully to the reference genome (99.03% mapping), and you have a relatively high number of duplicate reads that have been identified. These are typically removed during preprocessing (as you've done with the deduplication step).

### Are These Analysis-Ready Reads?

Yes, after deduplication and base recalibration, the reads are now considered **analysis-ready**. The high mapping percentage and processing steps (such as marking duplicates and base quality score recalibration) have prepared these reads for downstream analysis, like **variant discovery** (SNP and indel calling).

### Next Steps:

Following the GATK best practices workflow (as shown in the image), the next steps would be:

1.  **Variant Calling**: Use a variant caller like HaplotypeCaller or UnifiedGenotyper to identify SNPs and indels.
2.  **Joint Genotyping**: If working with multiple samples, joint genotyping can be performed to improve accuracy.
3.  **Filter Variants**: Once variants are called, filter them to remove low-quality or false positives.
4.  **Annotate Variants**: Add functional annotations to the variants (using tools like VEP or snpEff).
5.  **Evaluate Callset**: Assess whether the variants make sense biologically and technically (checking for artifacts, unexpected patterns, etc.).

After this, you will have a set of **high-quality variants** that can be used for downstream analyses.